src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -193,13 +204,29 @@ private ApiEvent InitializeApiEvent(string accountId)
             apiEvent.IsLegacyCacheEnabled = AuthenticationRequestParameters.RequestContext.ServiceBundle.Config.LegacyCacheCompatibilityEnabled;
             apiEvent.CacheInfo = CacheRefreshReason.NotApplicable;
             apiEvent.TokenType = AuthenticationRequestParameters.AuthenticationScheme.TelemetryTokenType;
+            apiEvent.AssertionType = GetAssertionType();
 
             // Give derived classes the ability to add or modify fields in the telemetry as needed.
             EnrichTelemetryApiEvent(apiEvent);
 
             return apiEvent;
         }
 
+        private AssertionType GetAssertionType()
+        {
+            if (ServiceBundle.Config.UseManagedIdentity)
+            {
+                return AssertionType.MSI;
+            }
+
+            if (AuthenticationRequestParameters.OnBeforeTokenRequestHandler != null)
+            {
+                return AssertionType.TokenRequestHandler;
```

Comment:
Let's not expose this one and just assume it's MSI. As it's only Azure SDK that need it.

Reply:
MSI will in the future be also needed by non-AzureSDK scenarios, or? For example, federated credentials

Reply:
We are exposing MSI via higher level APIs @rymeskar e.g. 

```csharp
ManagedIdentityApplication app = ManagedIdentityApplicationBuilder.Create(user_assigned_identity_client_id).Build();
var token = await app.AcquireToken("https://azure.keyvault.com/.default");
```

Afaik federated credentials are just another type of client assertion, we did not have to expose new API in MSAL.

```csharp
var cca = ConfidentialClientApplicationBuilder....WithClientAssertion(tokenFromOtherIdp).Build();
```

Reply:
These higher-Level APIs stay inside MSAL?

Reply:
Yes, these APIs will stay in MSAL.